RRTMGP refactoring

dustinswales opened the pull request:
This PR contains changes to the RRTMGP physics.

- Major refactoring of RRTMGP interstitials. collapse all rrtmgp_lw(sw) scheme files into rrtmgp_lw(sw)_main schemes. This reduces the number of Interstitial fields needed by the GP scheme and allows us to use the rte-rrtmgp submodule directly, w/o the need for a special branch containing the CCPP hooks.
- For longwave radiation calculation, add "blocking" capability to main loop.
- Fix bug in coupling of RRTMGP shortwave radiation to surface.
- CMake cleanup.

dustinswales committed:
Add loop over columns in RRTMGP longwave scheme. Collapse GP schemes into loop.

dustinswales committed:
Add loop over columns in RRTMGP scheme. Collapse GP schemes into loop. Removed deprecated scheme files.

dustinswales committed:
Bug fixes. Working now.

dustinswales committed:
Some more cleanup.

dustinswales committed:
Merge branch 'main' of https://github.com/NCAR/ccpp-physics into addNcolLoop2GP

dustinswales committed:
Longwave RRTMGP loop over ncol working.

dustinswales committed:
Merge branch 'addNcolLoop2GP' of https://github.com/dustinswales/ccpp-physics into HEAD

dustinswales committed:
Merge branch 'main' of https://github.com/NCAR/ccpp-physics into HEAD

dustinswales committed:
Changes for refactor

dustinswales committed:
Getting really close...

dustinswales committed:
Updated submodule pointer

dustinswales committed:
Some housekeeping. Working in UFS.

dustinswales committed:
Reorganize logic in solver loop

dustinswales committed:
Address issue 957

Commit:
Some modifications for rrtmgp physics blocking to work.

dustinswales committed:
Revert "Some modifications for rrtmgp physics blocking to work."

This reverts commit d35be37ed9441d449efeb3dae2c536ceb1cb4cf9.

dustinswales committed:
Move allocation of RRTMGP DDTs to init

Commit:
Merge branch 'ufs/dev' of https://github.com/ufs-community/ccpp-physics into HEAD

Commit:
Merge branch 'main' of https://github.com/NCAR/ccpp-physics into HEAD

Commit:
Commit before sync with upstream

Commit:
Merge branch 'ufs/dev' of https://github.com/ufs-community/ccpp-physics into HEAD

grantfirl commented on physics/GFS_rrtmgp_cloud_mp.F90:
```diff
@@ -1,10 +1,5 @@
-!! \brief This module contains the interface for ALL cloud microphysics assumptions and 
-!! the RRTMGP radiation scheme. Specific details below in subroutines.
-!!
+! ########################################################################################
```
I'm curious why this Doxygen markup was removed since the contents of this file don't really change much.

dustinswales replied:
Fixed.

grantfirl commented on physics/GFS_rrtmgp_post.F90:
```diff
@@ -0,0 +1,394 @@
+         sfcflw               ! LW radiation fluxes at sfc
+    type(sfcfsw_type), dimension(:), intent(inout) :: &
+         sfcfsw               ! SW radiation fluxes at sfc
+    type(topfsw_type), dimension(:), intent(out) :: &
```
I don't know if this will lead to problems or not, but topfsw and topflw are not set to anything if doLWrad and doSWrad are F since they are intent(out).

dustinswales replied:
They aren't used anywhere not guarded by doLWrad and doSWrad, but you're correct. I cleaned this up

grantfirl commented on physics/GFS_rrtmgp_pre.F90:
```diff
@@ -164,7 +166,8 @@ subroutine GFS_rrtmgp_pre_run(me, nCol, nLev, nTracers, i_o3, lsswr, lslwr, fhsw
     integer, intent(out) :: &  
```
Same comment RE: intent(out). Since many of these routines immediatley return if doSWrad and doLWrad are F, any intent(out) variables will automatically be set to random garbage upon return (whatever was previously in that memory space). This may not matter if they're not used, but it is good practice (and in the CCPP rules somewhere?) to initialize all intent(out) variables in CCPP schemes.

grantfirl commented on physics/GFS_rrtmgp_pre.F90:
```diff
@@ -15,16 +15,19 @@ module GFS_rrtmgp_pre
-  ! RRTMGP types
-  use mo_gas_concentrations, only: ty_gas_concs
   use radiation_tools,       only: check_error_msg,cmp_tlev
+  use rrtmgp_lw_gas_optics,  only: lw_gas_props  
```
@dustinswales This reminds me to ask about RRTMGP DDTs, since I've forgotten where this landed when it was implemented. In the past, I think that the we have said that if CCPP schemes use their own DDTs that they be confined to those that only used "internally" in physics. I.e., that in the long run we're not requiring hosts to "know" about more DDTs than necessary. So, in practice, that would mean confining RRTMGP-specific DDTs to the current Interstitial top-level DDT because all of those will technically be able to be allocated by the CCPP in the future with capgen. Is this your understanding and are your changes consistent with this? Am I totally off about this?

dustinswales replied:
@grantfirl This PR address the concerns you brought up. Up until now, there were two types of DDTs in the GP scheme: internal DDTs that exist with out the model's knowledge (i.e. lw_gas_props), and interstitial DDT's. For the former we had to add hooks for ccpp compliance in the rte-rrtmgp submodule, and also inform the host about these GP DDTs. This was a headache to manage and is now longer needed with the refactoring.

grantfirl commented on physics/GFS_rrtmgp_sw_pre.F90:
```diff
@@ -1,95 +0,0 @@
-       enddo
-
-       ! Spread across all SW bands
-       do iBand=1,sw_gas_props%get_nband()
```
Were these lines (sfc_alb_X_Y_byband) moved somewhere else or just removed?

dustinswales replied:
They were moved to [rrtmgp_sw_main.F90](https://github.com/dustinswales/ccpp-physics/blob/rrtmgp_refactor/physics/rrtmgp_sw_main.F90#L359). This is also part of the shortwave surface coupling bugfix (there is a similar spectral correction done in the _post step).

grantfirl commented on physics/rrtmgp_aerosol_optics.F90:
```diff
@@ -65,34 +57,37 @@ subroutine rrtmgp_aerosol_optics_run(doSWrad, doLWrad, nCol, nLev, nTracer, nTra
-         sw_optical_props_aerosol ! RRTMGP DDT: Longwave aerosol optical properties (tau)
-    type(ty_optical_props_1scl),intent(inout) :: &
-         lw_optical_props_aerosol ! RRTMGP DDT: Longwave aerosol optical properties (tau)
+    real(kind_phys), dimension(:,:,:), intent(out) :: &
```
Same comment about uninitialized intent(out) variables. It could be fine; just need to doublecheck. I mean, it wasn't causing problems before, so why should it now?

grantfirl commented on physics/rrtmgp_aerosol_optics.F90:
```diff
@@ -65,34 +57,37 @@ subroutine rrtmgp_aerosol_optics_run(doSWrad, doLWrad, nCol, nLev, nTracer, nTra
 
     ! Local variables
-    real(kind_phys), dimension(nCol, nLev, lw_gas_props%get_nband(), NF_AELW) :: &
+    real(kind_phys), dimension(nCol, nLev, lw_gas_props%get_nband(), 3) :: &
```
seems a little weird to be going from NF_AELW to a hard-coded constant. I guess it will never change?

dustinswales replied:
@grantfirl In this context 3 is fixed. The third dimension indexes different fields in a stacked array, optical-depth, single-scattering albedo, and asymmetry parameter

grantfirl commented on physics/rrtmgp_lw_cloud_optics.F90:
```diff
@@ -74,55 +72,42 @@ module rrtmgp_lw_cloud_optics
 
 contains
 
-!>\defgroup rrtmgp_lw_cloud_optics_mod GFS RRTMGP-LW Cloud Optics Module
```
Same comment about removing Doxygen stuff. We may need to consult Man about this to make sure it doesn't mess up the scientific docs too much.

grantfirl commented on physics/rrtmgp_lw_cloud_optics.F90:
```diff
@@ -391,171 +376,4 @@ subroutine rrtmgp_lw_cloud_optics_init(nrghice, mpicomm, mpirank, mpiroot,
-!! Compute longwave optical prperties (optical-depth) for ALL cloud types visible to RRTMGP.                                                                                                                                   
-!!
-!! \section rrtmgp_lw_gas_optics_run
-  subroutine rrtmgp_lw_cloud_optics_run(doLWrad, doG_cldoptics, icliq_lw, icice_lw,      &
```
I can't find where this code was moved to. Was it removed?

grantfirl replied:
rrtmgp_lw_main?

dustinswales replied:
Yes. The rrtmgp_lw_cloud_optics_run routine is now in [rtmgp_lw_main.F90](https://github.com/dustinswales/ccpp-physics/blob/rrtmgp_refactor/physics/rrtmgp_lw_main.F90#L362)
Similarly, the rrtmgp_lw_gas_optics_run routine is called just before the cloud-optics in the link above.

grantfirl commented on physics/rrtmgp_sw_main.F90:
```diff
@@ -0,0 +1,682 @@
+         errmsg                ! CCPP error message
+    integer, intent(out) :: &
+         errflg                ! CCPP error flag
+    real(kind_phys), dimension(:,:), intent(out) :: &
```
Same comment RE: intent(out)

grantfirl commented:
@dustinswales I'm finished looking through this. Looks fine to me for the most part, with some worries about uninitialized intent(out) variables. It seems like every small scheme that was deleted ended up in the SW or LW main schemes? It's a little difficult to follow code moving between files in the GitHub interface (without looking really hard).

dustinswales committed:
Added documenation back into scheme file.

dustinswales committed:
Address reviewers comments

dustinswales committed:
Address more reviewers comments

dustinswales commented:
> @dustinswales I'm finished looking through this. Looks fine to me for the most part, with some worries about uninitialized intent(out) variables. It seems like every small scheme that was deleted ended up in the SW or LW main schemes? It's a little difficult to follow code moving between files in the GitHub interface (without looking really hard).

@grantfirl Thanks for your review. I addressed your concerns about the intent of some variables.

I admit this is hard to follow, my apologizes, it's a total restructuring of how GP is implemented in the CCPP.

ChunxiZhang-NOAA commented on physics/GFS_rrtmgp_cloud_mp.F90:
```diff
@@ -484,47 +485,48 @@ subroutine cloud_mp_SAMF(nCol, nLev, t_lay, p_lev, p_lay, qs_lay, relhum,
-             clwc   = max(0.0, cnv_mixratio(iCol,iLay)) * con_g * deltaP
-             cld_cnv_iwp(iCol,iLay) = clwc * tem1
-             cld_cnv_lwp(iCol,iLay) = clwc - cld_cnv_iwp(iCol,iLay)
+             clwc   = max(0.0, cnv_mixratio(iCol,iLay)) * tem0 * deltaP
```
You modified from con_g to 1.0e5/con_g. What is 1.0e5?

dustinswales replied:
@ChunxiZhang-NOAA This was a bug fix identified by @Qingfu-Liu back in June.

ChunxiZhang-NOAA replied:
@dustinswales Thanks for clarifying.

jkbk2004 commented:
We are about to start working on https://github.com/ufs-community/ufs-weather-model/pull/1571. @dustinswales can you mark out unsolved conversations?

dustinswales committed:
Merge branch 'ufs/dev' of https://github.com/ufs-community/ccpp-physics into HEAD

dustinswales committed:
Interface changes for SP build in GP
